### PR TITLE
Demystifying OCaml Syntax with Merlin and OCaml-LSP

### PR DESCRIPTION
**Pull request by PizieDust**
# FEATURE: Add Syntax Documentation to Hover request

## Introduction
This PR introduces a new feature to OCaml-LSP for both beginners and also more experienced programmers. No more searching for obscure keywords or symbols on Google trying to find documentation. 
With this PR, you can hover over OCaml code in your editor, and instantly see a clear and summarized explanation of its syntax along with a link to the exact portion of OCaml's documentation for it.

### Benefits:
- Faster understanding: At a glance you can get a quick description of what the syntax is about.
- Improved learning: Beginners can easily understand how their code works while writing the code.
- Enhanced productivity: Very useful for understanding codebases and what a syntax is supposed to do.

### How it works:
- Hover your cursor over the OCaml code.
- The hover request will call the new `Syntax Documentation` command from Merlin which returns a `JSON` output containing a `name`, some `description` and a `url` pointing to the documentation.
- This information is then parsed and displayed on your editors hover UI.

### Scope
Currently the command covers only specific sections of OCaml's syntax such as `Language Extensions`. Subsequently, more syntax will be covered.

## Code Example
Example: Hovering over type definitions in VS code:
Code example:
```ocaml
(*  Variant type *)
type p = 
  | A (** comment for A*)
  | B (** comment for B*)

(* Record type *)
type point = {x: int; y: int}
```

## Current Behaviour
Before `Syntax document` command: UI will display only type definition 
> image

> image

Or display comments if they are available
> image

## New Feature
The UI now includes a short description about the syntax highlighted under the cursor.

> image

UI will also display comments if they are available 

> image

New `Syntax Document` command can differentiate different kinds of syntax and return appropriate documentation

> image

## WIP/Dependencies
This PR relies on a new Merlin command found in this PR (https://github.com/ocaml/merlin/pull/1706) which is currently unmerged. To test this feature out, add a pin dependency of Merlin to https://github.com/PizieDust/merlin/tree/syntax_documentation.

This PR is related to https://github.com/ocaml/merlin/issues/1692 

> @pitag-ha suggested using `JSON` instead of one blob string for the `Syntax Document` command which works great for this purpose.

cc @voodoos @pitag-ha

### CODE REVIEW

**Comment by jfeser** [comment]
I suspect that users will want this to be configurable. I could see how novice users might find this information helpful, but it's noisy. 

https://github.com/ocaml/ocaml-lsp/issues/1198 suggests go-to links for types. We could consider a go-to link for documentation, instead of inlining the documentation sentence into the hover.

**Comment by PizieDust** [comment]
> #1198 suggests go-to links for types. We could consider a go-to link for documentation, instead of inlining the documentation sentence into the hover.

This is certainly a good idea. I like the [Go to Documentation]() format. It definitely trims is down a bunch.

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:62]: You should not have to rely on such tests. Merlin commits to always give a manual page url so either there is no documentation at all, or the url should not be empty.

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:180]: There is no need to pass each field separately here. You can pass the whole record and destruct it later.

Also, there is sometimes no syntax documentation so you should pass a `Query.protocol.info option` or simply keep the `` `No_documentation / `Found _ `` values returned by merlin (as said in another comment).

**Review by voodoos — COMMENTED** [review]
Made a few comments, you are doing good progress !

To make the feature opt-in (or opt-out ?) can take inspiration from other configurable features of ocaml-lsp like "extended hover".

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:28]: This line will be reflect multiple other changes I suggested:
- We can call the argument `syntax_doc` here and everywhere else, it's shorter and mimic `doc`.
- There is no need to annotate the argument with it's type, especially when this type will simply be a `syntax_doc_result option`
- I think there it's too late to check if the feature is enabled: we don't want Merlin to perform extra work if it's not going to be displayed in the end.

File: ocaml-lsp-server/src/document.ml
Comment (voodoos) [inline ocaml-lsp-server/src/document.ml:320]: This means that Merlin will have to perform extra work that most of the time gets thrown away. We should not call `syntax_doc` if the option is not enabled and always return `false` in that case.

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:38]: The shorter, the better here: I think we can simply put the `See [Manual]` link at the end of the line without adding an extra separator.

We might also want to start the line with a constant prefix like `_syntax_ ` to make it clearer what the section is about ?

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:33]: As stated in other comments, checking that the feature is enabled should be done before we query Merlin. Then in this function we simply have to check if there is `Some syntax_doc` or `None`.

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:17]: This case is useless, `String.concat ~sep:"some sep" []` should already return `""`

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:57]: You don't need to call `sprintf` since `print_dividers` already returns a string (`sprintf "%s"` is always a useless call).

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:41]: This should be options, not empty strings for a more idiomatic approach.
Since `syntax_doc` will also be an option you will simply have to to an `Option.map`: 

```
Option.map syntax_doc ~f:(fun result -> sprintf "_syntax_ %s: %s. See [Manual](%s)" result....
```

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:49]: Same here, you can keep `doc` as an option and use `let doc = Option.map doc ...`.

Then at the end, to print all non-None info with the required dividers you can simply filter the list:

```
print_dividers (List.filter_opt [type_info; syntax_doc; doc])
```

File: ocaml-lsp-server/src/hover_req.ml
Comment (voodoos) [inline ocaml-lsp-server/src/hover_req.ml:24]: You're not formatting the code blocks the same as before (not sure if the extra spaces have an impact however).
But more importantly this function should have two arguments: the name and the type, which are both strings. No need for a list here given the usage of the function.

File: ocaml-lsp-server/docs/ocamllsp/config.md
Comment (PizieDust) [reply ocaml-lsp-server/docs/ocamllsp/config.md:28]: Okay great. I didn't if to use the current version or the version if a new release is made so I just left it blank. I'll update in the next commit.

File: ocaml-lsp-server/src/document.ml
Comment (PizieDust) [reply ocaml-lsp-server/src/document.ml:320]: Thanks for pointing this out

**Comment by smorimoto** [comment]
Changes for this PR on the VSCode extension side have just been merged.

File: ocaml-lsp-server/src/env_vars.ml
Comment (rgrinberg) [inline ocaml-lsp-server/src/env_vars.ml:21]: Is this config option experimental? If it's meant for end users, it shouldn't be an environment variable but a proper config option that can be passed when initializing the server.

Comment (PizieDust) [reply ocaml-lsp-server/src/env_vars.ml:21]: So, currently it works via the environment variable OR via a config option, whichever way the user chooses to activate it. 
I'd environment variables are not very convenient, i could easily refactor to just use only config options.

Comment (rgrinberg) [reply ocaml-lsp-server/src/env_vars.ml:21]: Yeah, I'd drop the env var. I mostly use them to ease debugging or for experimental stuff. Users already have the LSP config system to work with. That should be enough.

Comment (PizieDust) [reply ocaml-lsp-server/src/env_vars.ml]: Great. The latest commit fixes this.

File: ocaml-lsp-server/src/hover_req.ml
Comment (rgrinberg) [inline ocaml-lsp-server/src/hover_req.ml:17]: Since this API is private, there's no need to maintain backwards compatibility. Thus, optional arguments aren't particularly useful and just allow the callers not to think about the correct value for this parameter. Making it required will be better

File: ocaml-lsp-server/src/document.ml
Comment (rgrinberg) [inline ocaml-lsp-server/src/document.ml:286]: This query is available in only recent version of merlin, right? If so, an appropriate constraint should be added to the ocaml-lsp-server package (in dune-project).

**Review by rgrinberg — CHANGES_REQUESTED** [review]

**Review by rgrinberg — APPROVED** [review]

**Comment by rgrinberg** [comment]
I'll let @voodoos merge

**Comment by voodoos** [comment]
> I'll let @voodoos merge

ack, I will try to release Merlin so that we can properly run the testsuite before merging.

**Review by voodoos — COMMENTED** [review]
Sorry for the late review. Nice job Pizie, it looks like it's almost ready !

I think you changed the formatting of the on_hover responses slightly which breaks other tests. (The 5.0/5.1 CI only pass because it doesn't run the tests.) Did you run them on your machine ? Did they pass on your machine ?

For example, the following diff is clearly linked to your new printing of the dividers:
```
    - ---
    +
    +  --- 
    +
 ```
 
 There is no reason to change the results of other tests in this PR, unless the new feature is activated. You can run the `make test` target to run the test on your machine and check that the result is correct.

**Comment by PizieDust** [comment]
> Sorry for the late review. Nice job Pizie, it looks like it's almost ready !
> 
> I think you changed the formatting of the on_hover responses slightly which breaks other tests. (The 5.0/5.1 CI only pass because it doesn't run the tests.) Did you run them on your machine ? Did they pass on your machine ?
> 
> For example, the following diff is clearly linked to your new printing of the dividers:
> 
> ```
>     - ---
>     +
>     +  --- 
>     +
> ```
> 
> There is no reason to change the results of other tests in this PR, unless the new feature is activated. You can run the `make test` target to run the test on your machine and check that the result is correct.

Ahh yeah, I was using `dune build @runtest` and I think that's why I didn't notice that the tests were failing. It's all sorted now.

File: ocaml-lsp-server/src/hover_req.ml
Comment (PizieDust) [reply ocaml-lsp-server/src/hover_req.ml:31]: I think so too. There should be no reason using markdown syntax for clients which require only plain text.

**Comment by voodoos** [comment]
Thanks @PizieDust, I will merge once the checks finish.

**Pull request merged**
